flink-python/pyflink/table/table_descriptor.py

```diff
@@ -0,0 +1,507 @@
+            gateway.jvm.Thread.currentThread().getContextClassLoader().loadClass(strategy)
+                   .newInstance())
+        return self
+
+
+class Schema(Descriptor):
+    """
+    Describes a schema of a table.
+
+    ..note::
+        Field names are matched by the exact name by default (case sensitive).
+    """
+
+    def __init__(self):
+        gateway = get_gateway()
+        self._j_schema = gateway.jvm.Schema()
+        super(Schema, self).__init__(self._j_schema)
+
+    def field(self, field_name, field_type):
+        """
+        Adds a field with the field name and the data type or type string. Required.
+        This method can be called multiple times. The call order of this method defines
+        also the order of the fields in a row. Here is a document that introduces the type strings:
+        https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/connect.html#type-strings
+
+        :param field_name: The field name.
+        :param field_type: The data type or type string of the field.
+        :return: This schema object.
+        """
+        if isinstance(field_type, (str, unicode)):
+            self._j_schema = self._j_schema.field(field_name, field_type)
+        else:
+            self._j_schema = self._j_schema.field(field_name, to_java_type(field_type))
+        return self
+
+    def from_origin_field(self, origin_field_name):
+        """
+        Specifies the origin of the previously defined field. The origin field is defined by a
+        connector or format.
+
+        E.g. from_origin_field("myString", Types.STRING).from("CSV_MY_STRING")
+
+        ..note::
+            Field names are matched by the exact name by default (case sensitive).
+
+        :param origin_field_name: The origin field name.
+        :return: This schema object.
+        """
+        self._j_schema = get_method(self._j_schema, "from")(origin_field_name)
+        return self
+
+    def proctime(self):
+        """
+        Specifies the previously defined field as a processing-time attribute.
+
+        E.g. field("proctime", Types.SQL_TIMESTAMP).proctime()
+
+        :return: This schema object.
+        """
+        self._j_schema = self._j_schema.proctime()
+        return self
+
+    def rowtime(self, rowtime):
+        """
+        Specifies the previously defined field as an event-time attribute.
+
+        E.g. field("rowtime", Types.SQL_TIMESTAMP).rowtime(...)
+
+        :param rowtime: A :class:`RowTime`.
+        :return: This schema object.
+        """
+        self._j_schema = self._j_schema.rowtime(rowtime._j_rowtime)
+        return self
+
+
+class FormatDescriptor(Descriptor):
+    """
+    Describes the format of data.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, j_format_descriptor):
+        self._j_format_descriptor = j_format_descriptor
+        super(FormatDescriptor, self).__init__(self._j_format_descriptor)
+
+
+class OldCsv(FormatDescriptor):
```

Comment:
The reason that not adding schema() is that schema() method has a dependency on `TableSchema` class, which depends on a complete type system implemented at FLINK-12408. How about add this method and `TableSchema` after FLINK-12408 merged?

Reply:
Why `TableSchema` needs all the type? I appreciate if you can explain more detail ?

Reply:
Because some methods of TableSchema will return `TypeInformation`. Currently the type system in flink python API can only express few `TypeInformation` objects in Java, not all of them. The works in FLINK-12408 will change this situation. So IMO we'd better implement it after FLINK-12408 merged.

Reply:
The user does not need to know the `TypeInformation`, make sense to me, and can you create the follow-up JIRA?

Reply:
Of course. I have created the JIRA. Here is the link: https://issues.apache.org/jira/browse/FLINK-12588